GPU/GPUTracking/TRDTracking/GPUTRDTrackerComponent.cxx

```diff
@@ -359,22 +360,14 @@ int GPUTRDTrackerComponent::DoEvent(const AliHLTComponentEventData& evtData, con
   fTracker->Reset();
   fChain->mIOPtrs.nMergedTracks = tracksTPC.size();
   fChain->mIOPtrs.nTRDTracklets = nTrackletsTotal;
-  fChain->AllocateIOMemory();
+  fChain->mIOPtrs.nTRDTriggerRecords = 1;
+  //fChain->AllocateIOMemory();
   fRec->PrepareEvent();
   fRec->SetupGPUProcessor(fTracker, true);
 
-  // loop over all tracklets
-  for (int iTracklet = 0; iTracklet < nTrackletsTotal; ++iTracklet) {
-    if (!hasMCtracklets) {
-      if (fTracker->LoadTracklet(tracklets[iTracklet])) {
-        return -EINVAL;
-      }
-    } else {
-      if (fTracker->LoadTracklet(tracklets[iTracklet], trackletsMC[iTracklet].mLabel)) {
-        return -EINVAL;
-      }
-    }
-  }
+  std::sort(tracklets, tracklets + nTrackletsTotal);
```

Comment:
This is actually illegal. You are reinterpret_casting the const away in the ptr casts above, which you are not allowed to do:
```
    if (blocks[iBlock].fDataType == (kAliHLTDataTypeTrack | kAliHLTDataOriginITS) && fRequireITStrack) {
      itsData = (AliHLTTracksData*)blocks[iBlock].fPtr;
      fBenchmark.AddInput(blocks[iBlock].fSize);
    } else if (blocks[iBlock].fDataType == (AliHLTTPCDefinitions::TracksOuterDataType() | kAliHLTDataOriginTPC)) {
      tpcData = (AliHLTTracksData*)blocks[iBlock].fPtr;
      fBenchmark.AddInput(blocks[iBlock].fSize);
    } else if (blocks[iBlock].fDataType == (kAliHLTDataTypeTrackMC | kAliHLTDataOriginTPC)) {
      tpcDataMC = (AliHLTTrackMCData*)blocks[iBlock].fPtr;
      fBenchmark.AddInput(blocks[iBlock].fSize);
    } else if (blocks[iBlock].fDataType == (AliHLTTRDDefinitions::fgkTRDTrackletDataType)) {
      tracklets = reinterpret_cast<GPUTRDTrackletWord*>(blocks[iBlock].fPtr);
      nTrackletsTotal = blocks[iBlock].fSize / sizeof(GPUTRDTrackletWord);
      fBenchmark.AddInput(blocks[iBlock].fSize);
    } else if (blocks[iBlock].fDataType == (AliHLTTRDDefinitions::fgkTRDMCTrackletDataType)) {
      hasMCtracklets = true;
      trackletsMC = reinterpret_cast<GPUTRDTrackletLabels*>(blocks[iBlock].fPtr);
      nTrackletsTotalMC = blocks[iBlock].fSize / sizeof(GPUTRDTrackletLabels);
      fBenchmark.AddInput(blocks[iBlock].fSize);
```
All these should be `reinterpret_cast<const ...>`.
And then this sort here is not possible.
You will have to create a local copy of the tracklets inside the tracker component, and then you can sort tha tracklets inside the copy. But you must not alter the input buffer.

Reply:
Hm, with `reinterpret_cast<const ...>` the compiler is complaining about an invalid conversion:
/home/oschmidt/alice/sw/SOURCES/AliRoot/0_O2/0/GPU/GPUTracking/TRDTracking/GPUTRDTrackerComponent.cxx:292:21: error: invalid conversion from ‘const AliGPU::gpu::GPUTRDTrackletLabels*’ to ‘AliGPU::gpu::GPUTRDTrackletLabels*’ [-fpermissive]
  292 |       trackletsMC = reinterpret_cast<const GPUTRDTrackletLabels*>(blocks[iBlock].fPtr);

But I do create a local copy of the tracklets now and work with them instead of the input data block directly.

Reply:
> Hm, with `reinterpret_cast<const ...>` the compiler is complaining about an invalid conversion:
> /home/oschmidt/alice/sw/SOURCES/AliRoot/0_O2/0/GPU/GPUTracking/TRDTracking/GPUTRDTrackerComponent.cxx:292:21: error: invalid conversion from ‘const AliGPU::gpu::GPUTRDTrackletLabels*’ to ‘AliGPU::gpu::GPUTRDTrackletLabels*’ [-fpermissive]

This is because you have the local definitions without const in line 274:
```
  GPUTRDTrackletWord* tracklets = nullptr;
  GPUTRDTrackletLabels* trackletsMC = nullptr;
```
These should be const as well.

Reply:
Of course, yes. Thanks a lot, this is fixed now